### PR TITLE
Ensure @ConfigProperties validation works in native mode

### PR DESCRIPTION
**Pull request by geoand**

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java [hunk 1]
```diff
@@ -20,7 +20,7 @@
     public boolean reportErrorsAtRuntime;
 
     /**
-     * If debug symbols should be included
+     * If debug symbols should be includedPanacheBaseEntity
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java]: You don't want that :).

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java [hunk 2]
```diff
@@ -127,12 +124,21 @@ AutoInjectAnnotationBuildItem autoInjectConfigProperty() {
...
     void validateConfigProperties(ConfigRecorder recorder, List<ConfigPropertyBuildItem> configProperties,
-            BeanContainerBuildItem beanContainer) {
+            BeanContainerBuildItem beanContainer, BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
         // IMPL NOTE: we do depend on BeanContainerBuildItem to make sure that the BeanDeploymentValidator finished its processing
 
+        // the non-primitive types need to be registered for reflection since Class.forName is used at runtime to load the class
+        for (ConfigPropertyBuildItem item : configProperties) {
+            Type requiredType = item.getPropertyType();
+            String propertyType = requiredType.name().toString();
+            if (requiredType.kind() != Kind.ARRAY && requiredType.kind() != Kind.PRIMITIVE) {
```
Comment (gsmet) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java]: If it's an array, you don't need to register the array element type?

**Review by gsmet — COMMENTED** [review]
Added a question.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java (hunk 1 above)
Comment (gsmet) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java]: I saw you fixed that one.

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java (hunk 2 above)
Comment (geoand) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java]: I just moved what already existed from the vanilla `@ConfigProperty` basically, but I do see your point. On the other hand we haven't had any reports of it failing so I would say it's probably OK.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java (hunk 1 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java]: yeah, it wasn't right by any means :)

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java (hunk 2 above)
Comment (gsmet) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java]: Or nobody uses arrays :).

**Review by gsmet — APPROVED** [review]

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java (hunk 2 above)
Comment (geoand) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java]: :laughing:

**Review by aloubyansky — APPROVED** [review]
This did fix the issue.

**Comment by gsmet** [comment]
Merged, thanks!

**Comment by geoand** [comment]
Thank you!

**Comment by mkouba** [comment]
@geoand Looks good. So IIUIC the problem was that for `@ConfigProperties` no `ReflectiveClassBuildItem` was produced..?

**Comment by geoand** [comment]
Exactly yes

**Pull request merged**